src/compiler/checker.ts

```diff
@@ -19701,11 +19702,18 @@ namespace ts {
             return declaredType;
         }
 
-        function getTypeFactsOfTypes(types: Type[]): TypeFacts {
+        function getTypeFactsOfTypes(types: Type[], isUnion: boolean): TypeFacts {
             let result: TypeFacts = TypeFacts.None;
             for (const t of types) {
                 result |= getTypeFacts(t);
             }
+            if (!isUnion) {
+                // Get the set of positive facts for the intersection by masking with the negative set shifted left,
+                // then shift those present positive facts into the negative fact value range, and unset any of those
+                // bits (by negating that mask and then intersecting it with the original value)
+                const positiveFacts = result & (TypeFacts.NegativeTypeofFacts >> 8);

```

Comment:
It seems like the test coverage here is not adequate to ensure that this math continues to work for all cases in the future. Let's have a testcase that makes sure that e.g. someone reordering the EQ members without also pairwise reordering the NE members sees a baseline diff.

Reply:
I'm not quite sure how to make a baseline for that, but a unittest can be made if the `TypeFacts` gets `export`  + `@internal`:

```ts
describe("getTypeFactsOfTypes", () => {
    // See: https://github.com/microsoft/TypeScript/pull/39016
    it("correctly can strip NegativeTypeofFacts when there are PositiveTypeofFacts of the same type via bitmask logic", () => {
        const TypeFacts = (ts as any).TypeFacts;

        // For a set of facts, which include both the positive and negative of each other
        let ExampleFacts = TypeFacts.TypeofEQNumber | TypeFacts.TypeofNENumber | TypeFacts.TypeofNEBigInt;
        // Using this line of code (with the 8 which effectively represents the number of EQ types)
        const positiveFacts = ExampleFacts & (TypeFacts.NegativeTypeofFacts >> 8);
        // Then this should drop the negative version of the same type
        ExampleFacts &= ~(positiveFacts << 8);
        assert.isTrue((ExampleFacts & TypeFacts.TypeofNENumber) === 0);

        // And it keeps the rest
        assert.isTrue((ExampleFacts & TypeFacts.TypeofEQNumber) !== 0);
        assert.isTrue((ExampleFacts & TypeFacts.TypeofNEBigInt) !== 0);

        // If this test has failed, you _probably_ need to adjust the 8 here, and in
        // getTypeFactsOfTypes.
    });
});```